File: workshops/numeric-computing/exercises/native-math/fixtures/pow/runner.jl
#!/usr/bin/env julia

import JSON

"""
    gen( b, x, name )

Generate fixture data and write to file.

# Arguments

* `b`: bases
* `x`: exponents
* `name::AbstractString`: output filename

# Examples

``` julia
julia> b = rand( 1001 )*10.0;
julia> x = rand( 1001 )*616.0 - 308.0;
julia> gen( b, x, \"data.json\" );
```
"""
function gen( b, x, name )
	y = Array( Float64, length(b) );
	for i in eachindex(b)
		y[ i ] = b[i]^x[i];
	end

	# Store data to be written to file as a collection:
	data = Dict([
		("b", b),
		("x", x),
		("expected", y)
	]);

	# Based on the script directory, create an output filepath:
	filepath = joinpath( dir, name );

	# Write the data to the output filepath as JSON:
	outfile = open( filepath, "w" );
	write( outfile, JSON.json(data) );
	close( outfile );
end


# Get the filename:
file = @__FILE__;

# Extract the directory in which this file resides:
dir = dirname( file );

# Generate fixture data:
b = sort( rand( 1001 )*10.0 );
x = rand( 1001 )*616.0 - 308.0;
gen( b, x, "data.json" );

File: tools/snippets/test/fixtures/r/runner.R
#!/usr/bin/env Rscript
#
# @license Apache-2.0
#
# Copyright (c) 2025 The Stdlib Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Set the precision to 16 digits:
options( digits = 16 );

#' Generate test fixtures.
#'
#' @examples
#' main();
main <- function() {
	#' Get the script filepath.
	#'
	#' @return The absolute path of this script
	#'
	#' @examples
	#' filepath <- get_script_path();
	get_script_path <- function() {
		args <- commandArgs( trailingOnly = FALSE );
		needle <- "--file=";
		match <- grep( needle, args );
		if ( length( match ) > 0 ) {
			# Rscript:
			filepath <- sub( needle, "", args[match] );
		} else {
			ls_vars <- ls( sys.frames()[[1]] )
			if ( "fileName" %in% ls_vars ) {
				# Source'd via RStudio:
				filepath <- sys.frames()[[1]]$fileName; # nolint
			} else {
				# Source'd via R console:
				filepath <- sys.frames()[[1]]$ofile;
			}
		}
		return( normalizePath( filepath ) );
	}

	#' Convert a data structure to JSON.
	#'
	#' @param x A data structure to convert
	#' @return JSON blob
	#'
	#' @examples
	#' x <- seq( -6.5, 25, 0.5 );
	#' json <- to_json( x );
	to_json <- function( x ) {
		return( jsonlite::toJSON( x, digits = 16, auto_unbox = TRUE ) );
	}

	#' Generate an output absolute filepath based on the script directory.
	#'
	#' @param name An output filename
	#' @return An absolute filepath
	#'
	#' @examples
	#' filepath <- get_filepath( "data.json" );
	get_filepath <- function( name ) {
		return( paste( source_dir, "/", name, sep = "" ) );
	}

	# Get the directory of this script:
	source_dir <- dirname( get_script_path() );

	# TODO: generate test fixture data (`x` and `y`), making sure to handle NaNs


	# Convert fixture data to JSON:
	x <- to_json( x );
	y <- to_json( y );

	# Write the data to file...
	filepath <- get_filepath( "data.json" );
	write( x, filepath );

	filepath <- get_filepath( "expected.json" );
	write( y, filepath );
}

main();
